cmd/route-advertisement-daemon/main.go

```diff
@@ -0,0 +1,359 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"net"
+	"net/netip"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/mdlayher/ndp"
+	"golang.org/x/net/ipv6"
+)
+
+var (
+	logFile       string
+	iface         string
+	router        string
+	isRemoteRoute bool
+	client        string
+	clientHWAddr  string
+	prefix        string
+
+	linkLocalAllRouters = netip.MustParseAddr("ff02::2")
+)
+
+func main() {
+	file, err := setRADLog()
+	if err != nil {
+		log.Fatalf("set route advertisement daemon log: %s", err)
+	}
+	defer file.Close()
+
+	src, dst, cidr, err := validateVars()
+	if err != nil {
+		log.Fatalf("validate vars: %s", err)
+	}
+	if src != nil && isRemoteRoute {
+		if _, err := executeCommand("ip", "-6", "neigh", "add", client, "lladdr", clientHWAddr, "dev", iface); err != nil {
+			log.Fatalf("add neighbor entry for client: %s", err)
+		}
+
+		ipv6LLA := src
+		if !src.IsLinkLocalUnicast() {
+			mac, err := tryDiscoveryNeighborMAC(iface, src, 5)
+			if err != nil {
+				log.Fatalf("discovery router MAC: %s", err)
+			}
+			ipv6LLA = generateEUI64IPv6Address(net.ParseIP("fe80::0"), mac)
+			mac2, err := tryDiscoveryNeighborMAC(iface, ipv6LLA, 5)
+			if err != nil {
+				log.Fatalf("discovery router MAC: %s", err)
+			}
+			if mac.String() != mac2.String() {
+				log.Fatalf("failed to get router link-local address")
+			}
+		}
+
+		if _, err := executeCommand("ip6tables", "-A", "OUTPUT", "-o", iface, "--src", ipv6LLA.String(), "-p", "icmpv6", "--icmpv6-type", "neighbor-solicitation", "-j", "DROP"); err != nil {
+			log.Fatalf("drop neighbor solicitation on interface: %s", err)
+		}
+		if _, err := executeCommand("ip6tables", "-A", "OUTPUT", "-o", iface, "--src", ipv6LLA.String(), "-p", "icmpv6", "--icmpv6-type", "neighbor-advertisement", "-j", "DROP"); err != nil {
+			log.Fatalf("drop neighbor advertisement on interface: %s", err)
+		}
+
+		// As described in RFC 4861 section-4.2, the srouce address of RA must be the link-local
+		// address assigned to the interface from which the message is sent, so the LLA of default
+		// router have to be added to the interface. And the followings need to be done.
+		//  1.Disable DAD of the interface
+		//  2.Add static neighbor entry for the client, otherwise the interface will send a NS
+		//    message with it's MAC in options to client
+		//  3.Drop NA message from interface responsed to NS message learning default router LLA
+		//  4.Drop NS message from interface with default router LLA in options
+		if executeCommand("ip", "addr", "add", fmt.Sprintf("%s/64", ipv6LLA.String()), "dev", iface); err != nil {
+			log.Fatalf("add IPv6 addr to the interface: %s", err)
+		}
+
+		src = ipv6LLA
+	}
+
+	if err := startRouteAdvertisement(iface, src, dst, cidr); err != nil {
+		log.Fatalf("start route advertisement: %s", err)
+	}
+}
+
+func setRADLog() (*os.File, error) {
```

Comment:
Maybe it's better to use a log lib that supports file truncate?

Reply:
Ok，i'll  use a lib to deal the log message.